### ISSUE TITLE
Rule proposal: no-useless-coercion

### ISSUE THREAD

**Issue by kaicataldo** [eslint/eslint#7545]
**Please describe what the rule should do:**
Refs https://github.com/eslint/eslint/issues/7310#issuecomment-253091809

`no-useless-coercion` would disallow type coercion using `Object()`, `Boolean()`, `String(),` and `Number()` if the argument is already of that type.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[ ] Warns about a potential error
[X] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

Examples of **incorrect** code for this rule:
```js
/*eslint no-useless-coercion: "error"*/

Object({a: 1});
Boolean(true);
String('hello world');
Number(12);
```

Examples of **correct** code for this rule:
```js
/*eslint no-useless-coercion: "error"*/

Object('foo');
Boolean(1);
String(123);
Number('hello world');
```

**Why should this rule be included in ESLint (instead of a plugin)?**
In the same spirit as `no-useless-concat`, `no-useless-escape`, `	
no-useless-return`, `no-useless-computed-key`, and `no-useless-rename`, this rule disallows something that has no effect whatsoever. If someone is writing code this way, they've either made a mistake or don't understand the use case for those methods, and should be warned.

**Comment by not-an-aardvark** [comment]
:+1:, but note that this overlaps slightly with [no-extra-boolean-cast](http://eslint.org/docs/rules/no-extra-boolean-cast) in that both of them disallow unnecessary coercions.

**Comment by kaicataldo** [comment]
Thanks for pointing that out - that's a fair point. The intention `no-extra-boolean-cast` looks to be to avoid multiple boolean casting, so I think it actually aims to solve a different issue (though I could see it being confusing to users what that difference is). It only checks for `Boolean(!!foo)` (casting once with `!!` and then a second time with `Boolean()`), and doesn't actually warn for `Boolean(true)`, so there won't be multiple errors reported for the case outlined in this proposal. Would love to hear others' thoughts on this too, though.

**Comment by alberto** [comment]
Not sure this is a common enough issue to be included in core. This kind of mistakes seems marginal to me and it's consequences are not dangerous.

**Comment by tbassetto** [comment]
Sure, it's not dangerous, but it wouldn't be the first rule in core preventing users from writing unnecessary code. I have found code like this, mainly in JavaScript codebase written by developers who just switched from strongly typed languages.

**Comment by kaicataldo** [comment]
@alberto For some [more context](https://github.com/eslint/eslint/issues/7310) on why this rule was suggested: Are you :-1: for this proposal?

**Comment by JamesHenry** [comment]
I would echo @not-an-aardvark's thoughts, am 👍 for the proposal but would be ideal if the overlap with `no-extra-boolean-cast` could be addressed in some way

**Comment by kaicataldo** [comment]
I'm closing this issues because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.
